### PR TITLE
Add CAN intr flags to config

### PR DESCRIPTION
**Pull request by elotom**
I run into an issue that is described here: [Running out of interrupts](https://esp32developer.com/programming-in-c-c/interrupts/running-out-of-interrupts). With this pull request, I would like to add interrupt flags to the config then I can use custom interrupt flags for the CAN driver.

### CODE REVIEW

**Comment by elotom** [comment]
I think CI failure is not related to my changes.

**Comment by MabezDev** [comment]
> I think CI failure is not related to my changes.

Correct, it should be fixed upstream once https://github.com/rust-lang/rust/pull/111234 is merged.

**Review by Vollbrecht — CHANGES_REQUESTED** [review]

File: src/can.rs [hunk 1]
```diff
@@ -483,6 +485,11 @@ pub mod config {
             self.alerts = alerts;
             self
         }
+
+        pub fn intr_flags(mut self, flags: i32) -> Self {
```
[the remark this answers is not in the document]
Comment (elotom) [reply src/can.rs]: I have added the IntrFlags enum

File: src/can.rs [hunk 2]
```diff
@@ -163,6 +163,67 @@ pub mod config {
...
+        // Accept a Level 6 interrupt vector.
+        Level6,
+        // Accept a Level 7 interrupt vector (highest priority)
+        Nmi,
+        // Interrupt can be shared between ISRs.
+        Shared,
+        // Edge-triggered interrupt.
+        Edge,
+        // ISR can be called if cache is disabled.
+        Iram,
```
Comment (Vollbrecht) [inline src/can.rs:189]: when this option is used it needs SDKCONFIG option CONFIG_TWAI_ISR_IN_IRAM enabled, so maybe gate it or add an info.

File: src/can.rs [hunk 3]
```diff
@@ -163,6 +163,67 @@ pub mod config {
         }
     }
 
+    /// Interrupt allocation flags.
+    /// These flags can be used to specify which interrupt qualities the code calling esp_intr_alloc* needs.
+    #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+    pub enum IntrFlags {
```
Comment (Vollbrecht) [inline src/can.rs:169]: did you check if all of the flags are present on all platforms? e.g on esp32, esp32s2 ,esp32s3,  esp32c3

**Review by Vollbrecht — CHANGES_REQUESTED** [review]
Thanks for the quick update! I have two additional small notes, otherwise it looks good! If you can elaborate on it, i think we are good to go.

File: src/can.rs (hunk 3 above)
Comment (elotom) [reply src/can.rs:169]: Good point, I have double-checked the documentation and it looks like those flags are present on all platforms.

File: src/can.rs (hunk 2 above)
Comment (elotom) [reply src/can.rs:190]: I have updated the comment here. I'm not sure if can.rs is a good place for those flags. They are not only related to TWAI peripherals for example [I2C](https://github.com/esp-rs/esp-idf-hal/blob/331f3392a44f9e4bef7935df4e7186376b89d2f6/src/i2c.rs#LL188C29-L188C29).

Comment (elotom) [reply src/can.rs]: As we discussed yesterday I have moved the flags to intterupt.rs which is definitely a better place for them.

**Review by Vollbrecht — APPROVED** [review]
Technically i see no problem here anymore, @ivmarkov  do you see a reason or have a long term plan why we shouldn't add the flags to our public api? Otherwise its a good to go from me.

**Comment by svenstaro** [comment]
@ivmarkov You had some comments in Matrix. Could you post them here again?

**Comment by ivmarkov** [comment]
I'll implement something very similar to what is proposed here, with some minor tweaks, and putting `unsafe` here and there.

@elotom @svenstaro Is "Level 1" actually the sane default for TWAI's interrupt alloc flags? Shouldn't it be 0, as all other drivers do (I2C, CAN, GPIO, etc.)?

**Comment by ivmarkov** [comment]
Should be addressed since https://github.com/esp-rs/esp-idf-hal/commit/ba9cc4246a18880cdcea5aea21ec3a23edd220f6

Note that the new default interrupt flags for TWAI are "0", as all other drivers use. If you think this is inappropriate, let me know why.

P.S. Turns out ESP IDF does a pretty thorough inspection of the passed interrupt flags, so putting `unsafe` was not necessary.

**Comment by elotom** [comment]
> I'll implement something very similar to what is proposed here, with some minor tweaks, and putting `unsafe` here and there.
> 
> @elotom @svenstaro Is "Level 1" actually the sane default for TWAI's interrupt alloc flags? Shouldn't it be 0, as all other drivers do (I2C, CAN, GPIO, etc.)?

In my opinion, to have better control over what level was allocated we should set it explicitly to level 1 (lowest priority) for all drivers and add API for changing that if needed. According to [docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/intr_alloc.html#_CPPv414esp_intr_allocii14intr_handler_tPvP13intr_handle_t) if intr_flag is 0 levels can be 1,2 or 3.
```flags – An ORred mask of the ESP_INTR_FLAG_* defines. These restrict the choice of interrupts that this routine can choose from. If this value is 0, it will default to allocating a non-shared interrupt of level 1, 2 or 3. If this is ESP_INTR_FLAG_SHARED, it will allocate a shared interrupt of level 1. Setting ESP_INTR_FLAG_INTRDISABLED will return from this function with the interrupt disabled.```

**Pull request closed without merging**